### PR TITLE
Remove Extension Marketplace and add details on Extension Settings

### PR DESCRIPTION
**Pull request by SBalslev**
Add a few fields to Extension details when access from Extension Management.
Remove the Extension Marketplace functionality (obsoleted in 24), hide it from the Tell me.
Add overview of the extensions that has source available for down on the extension list.
Add overview of resource protection policies defined by the publisher on the extension details page.

Fixes [AB#537523](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/537523)

### CODE REVIEW

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Page.al [hunk 1]
```diff
@@ -1,116 +0,0 @@
...
-{
-    Caption = 'Extension Marketplace';
-    AdditionalSearchTerms = 'app,add-in,customize,plug-in,appsource';
-    PageType = Card;
-    ApplicationArea = All;
-    UsageCategory = Administration;
-    Editable = false;
-    ObsoleteState = Pending;
-    ObsoleteReason = 'This page will be obsoleted. Microsoft AppSource apps feature will replace the Extension Marketplace.';
-    ObsoleteTag = '24.0';
```
Comment (grobyns) [inline src/System Application/App/Extension Management/src/ExtensionMarketplace.Page.al]: ehh... it seems like we're not respecting the obsoletion period here. If something was obsoleted in v24 we should keep it functional until v26 at least and remove in v27 at the earliest, and since i see preprocessor symbols CLEAN24 added everywhere that process should be automatic. 
Removing it earlier would be seen as an unexpected breaking change.

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al [hunk 2]
```diff
@@ -487,12 +487,6 @@ codeunit 2501 "Extension Marketplace"
...
     local procedure OpenAppSourceMarket()
     begin
-#if not CLEAN24        
-#pragma warning disable AL0432
-        Page.Run(Page::"Extension Marketplace");
-#pragma warning restore AL0432
-#else        
-        Hyperlink('https://aka.ms/BCAppSource');
-#endif        
+        Page.Run(2515);
```
Comment (StefanSosic) [inline src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al]: Why hardcode with number? 
Much more readable with Page::"AppSource Product List"

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al
Comment (StefanSosic) [inline src/System Application/App/Extension Management/src/ExtensionManagement.Page.al:1]: Can we also change in this file Page.Run(2515); to be Page::"AppSource Product List"?

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Page.al (hunk 1 above)
Comment (StefanSosic) [reply src/System Application/App/Extension Management/src/ExtensionMarketplace.Page.al]: Yeah, I would agree here.
In order to be introduced now, we could probably hide old one

**Review by StefanSosic — CHANGES_REQUESTED** [review]

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al (hunk 2 above)
Comment (blrobl) [reply src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al]: This is because Extension Management is a dependency of AppSource Apps module. If we were to do this explicit call then we would need to add Appsource Apps as a dependency of Extension Management, leading to a circular dependency. This is a way to bypass that :)

Comment (StefanSosic) [reply src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al]: True and agree. Missed that, of course not add circular dependency only for that :) 

It's just that you don't see which Page it is with hovering and Ctrl+Click also doesn't open it up, maybe adding a comment saying Page "xxx" in Module "xxx" for quick sight, without opening symbols and checking there. Just an idea.

File: src/System Application/App/Extension Management/src/ExtensionInstallationImpl.Codeunit.al [hunk 3]
```diff
@@ -364,5 +364,41 @@ codeunit 2500 "Extension Installation Impl"
...
+    end;
+
+    internal procedure AllowsDownloadSourceInSymbols(ResourceProtectionPolicy: Integer): Boolean
+    begin
+        exit(HasPolicy(ResourceProtectionPolicy, 4));
+    end;
+
+    local procedure HasPolicy(ResourceProtectionPolicy: Integer; PolicyToTestFor: Integer): Boolean
+    var
+        remaining, tester, result, nextFlag : integer;
```
Comment (StefanSosic) [inline src/System Application/App/Extension Management/src/ExtensionInstallationImpl.Codeunit.al]: Variable names casing, also integer -> Integer

File: src/System Application/App/Extension Management/src/ExtensionSettings.Page.al [hunk 4]
```diff
@@ -92,10 +136,16 @@ page 2511 "Extension Settings"
...
         AppNameValue: Text;
         AppPublisherValue: Text;
         AppIdValue: Text;
+        AppVersionDisplay: Text;
+        AppIsInstalled: Boolean;
+        PublishedAs: Text;
+        IsTenantExtension: Boolean;
+        AppAllowsDebuggging: Boolean;
+        AppAllowsDownloadSource: Boolean;
+        AppAllowsDownloadSourceInSymbols: Boolean;
```
Comment (StefanSosic) [inline src/System Application/App/Extension Management/src/ExtensionSettings.Page.al]: Sorting by variable types

File: src/System Application/App/Extension Management/src/ExtensionInstallationImpl.Codeunit.al [hunk 5]
```diff
@@ -364,5 +364,44 @@ codeunit 2500 "Extension Installation Impl"
...
+    begin
+        Result := 0;
+        Remaining := ResourceProtectionPolicy;
+        Tester := PolicyToTestFor;
+        NextFlag := 1;
+        while ((Remaining > 0) and (Tester > 0)) do begin
+            if ((Remaining mod 2) = 1) and ((Tester mod 2) = 1) then
+                result += nextFlag;
+
+            remaining := Remaining div 2;
```
Comment (NKarolak) [inline src/System Application/App/Extension Management/src/ExtensionInstallationImpl.Codeunit.al]:
```suggestion
        while (Remaining > 0) and (Tester > 0) do begin
            if ((Remaining mod 2) = 1) and ((Tester mod 2) = 1) then
                Result += NextFlag;

            Remaining := Remaining div 2;
```

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Page.al (hunk 1 above)
Comment (SBalslev) [reply src/System Application/App/Extension Management/src/ExtensionMarketplace.Page.al]: You're right we were too eager to remove it with the deprecation from the app source team. Left the page here but removed the functionality. Should not break anyone except from a bad experience if they try to use it (blank page with only one new action).

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al [hunk 6]
```diff
@@ -64,6 +64,14 @@ page 2500 "Extension Management"
                     StyleExpr = InfoStyle;
                     ToolTip = 'Specifies whether the extension is installed.';
                 }
+                field("Source"; AllowsDownloadSource)
+                {
+                    Caption = 'Source';
+                    StyleExpr = AllowsDownloadSourceStyleExpr;
+                    ToolTip = 'Specifies whether the extension is allows the source to be downloaded.';
```
Comment (NKarolak) [inline src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]:
```suggestion
                    ToolTip = 'Specifies whether the extension allows the source to be downloaded.';
```

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al [hunk 7]
```diff
@@ -208,37 +216,15 @@ page 2500 "Extension Management"
...
-#endif
-                }
                 action("Microsoft AppSource Gallery")
                 {
                     Caption = 'AppSource Gallery';
                     Enabled = IsSaaS;
                     Image = NewItem;
                     ToolTip = 'Browse the Microsoft AppSource Gallery for new extensions to install.';
                     Visible = not IsOnPremDisplay;
+                    RunObject = Page 2515;
```
Comment (NKarolak) [inline src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]: What about adding a comment here to explain why the page ID instead of the name is used? See also
https://github.com/microsoft/BCApps/pull/1435/files#r1660628195

File: src/System Application/App/Extension Management/src/ExtensionInstallationImpl.Codeunit.al (hunk 3 above)
Comment (SBalslev) [reply src/System Application/App/Extension Management/src/ExtensionInstallationImpl.Codeunit.al]: remains one of the hardest things about AL code

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al [hunk 8]
```diff
@@ -366,11 +353,15 @@ page 2500 "Extension Management"
         // Determining Record and Styling Configurations
         IsInstalled := ExtensionInstallationImpl.IsInstalledByPackageId(Rec."Package ID");
         IsTenantExtension := Rec."Published As" <> Rec."Published As"::Global;
-    end;
-
-    local procedure GetVersionDisplayText(): Text
-    begin
-        exit(StrSubstNo(VersionFormatTxt, ExtensionInstallationImpl.GetVersionDisplayString(Rec)));
+        AllowsDownloadSource := AllowsDownloadSource::" ";
+        if (IsTenantExtension) then
```
Comment (NKarolak) [inline src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]:
```suggestion
        if IsTenantExtension then
```

File: src/System Application/App/Extension Management/src/ExtensionSettings.Page.al [hunk 9]
```diff
@@ -30,37 +31,75 @@ page 2511 "Extension Settings"
...
+            group("Resource protection policies")
+            {
+                Visible = IsTenantExtension;
+
+                Caption = 'Resource Protection Policies defined by the extension';
+                field(AppAllowsDebuggging; AppAllowsDebuggging)
+                {
+                    Caption = 'Allow Debugging';
+                    Editable = false;
+                    ToolTip = 'Specifies whether the publisher permits the runtime debug this extension .';
```
Comment (NKarolak) [inline src/System Application/App/Extension Management/src/ExtensionSettings.Page.al]:
```suggestion
                    ToolTip = 'Specifies whether the publisher permits the runtime to debug this extension .';
```

**Review by NKarolak — CHANGES_REQUESTED** [review]

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al (hunk 7 above)
Comment (SBalslev) [reply src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]: Not a fan but there we go :)

File: src/System Application/App/Extension Management/src/ExtensionSettings.Page.al (hunk 9 above)
Comment (SBalslev) [reply src/System Application/App/Extension Management/src/ExtensionSettings.Page.al]: Thanks :)

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al (hunk 7 above)
Comment (NKarolak) [reply src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]: @SBalslev Resolved but still unchanged? ;-)
```suggestion
                    RunObject = Page 2515; // Using ID to avoid circular dependency
```

Comment (SBalslev) [reply src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]: wierd - let me try again

Comment (SBalslev) [reply src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]: One of 3 got lost somewhere - it's back now.

Comment (NKarolak) [reply src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]: I've seen this happening over and over again - requests are marked as "Resolved", but the code was not updated in fact.
Maybe some GitHub glitch that should be paid attention to?

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al [hunk 10]
```diff
@@ -451,11 +450,14 @@ codeunit 2501 "Extension Marketplace"
         exit(false);
     end;
 
+#if not CLEAN24
+    [Obsolete('Microsoft AppSource apps feature will replace the Extension Marketplace.', '24.0')]
```
Comment (NKarolak) [inline src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al]: Not as per current app.json?

```suggestion
#if not CLEAN25
    [Obsolete('Microsoft AppSource apps feature will replace the Extension Marketplace.', '25.0')]
```

Comment (SBalslev) [reply src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al]: ha you're right - I took the one from the marketplace page

File: src/System Application/App/Extension Management/src/ExtensionInstallationImpl.Codeunit.al [hunk 11]
```diff
@@ -364,5 +364,44 @@ codeunit 2500 "Extension Installation Impl"
...
+
+    internal procedure AllowsDownloadSource(ResourceProtectionPolicy: Integer): Boolean
+    begin
+        exit(HasPolicy(ResourceProtectionPolicy, 2));
+    end;
+
+    internal procedure AllowsDownloadSourceInSymbols(ResourceProtectionPolicy: Integer): Boolean
+    begin
+        exit(HasPolicy(ResourceProtectionPolicy, 4));
+    end;
```
Comment (NKarolak) [inline src/System Application/App/Extension Management/src/ExtensionInstallationImpl.Codeunit.al]: The codeunit already is internal, and the other procedures are also defined as global.

```suggestion
    procedure AllowsDebug(ResourceProtectionPolicy: Integer): Boolean
    begin
        exit(HasPolicy(ResourceProtectionPolicy, 1));
    end;

    procedure AllowsDownloadSource(ResourceProtectionPolicy: Integer): Boolean
    begin
        exit(HasPolicy(ResourceProtectionPolicy, 2));
    end;

    procedure AllowsDownloadSourceInSymbols(ResourceProtectionPolicy: Integer): Boolean
    begin
        exit(HasPolicy(ResourceProtectionPolicy, 4));
    end;
```

**Review by NKarolak — CHANGES_REQUESTED** [review]

File: src/System Application/App/Extension Management/src/ExtensionInstallationImpl.Codeunit.al (hunk 11 above)
Comment (SBalslev) [reply src/System Application/App/Extension Management/src/ExtensionInstallationImpl.Codeunit.al]: I like being explicit about accessibility but fair point.

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al [hunk 12]
```diff
@@ -64,6 +64,14 @@ page 2500 "Extension Management"
...
                     ToolTip = 'Specifies whether the extension is installed.';
                 }
+                field("Source"; AllowsDownloadSource)
+                {
+                    Caption = 'Source';
+                    StyleExpr = AllowsDownloadSourceStyleExpr;
+                    ToolTip = 'Specifies whether the extension allows the source to be downloaded.';
+                    OptionCaption = ' ,Yes,No';
+                }
+
```
Comment (NKarolak) [inline src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]:
```suggestion
```

File: src/System Application/App/Extension Management/src/ExtensionSettings.Page.al [hunk 13]
```diff
@@ -30,37 +31,75 @@ page 2511 "Extension Settings"
...
                     Editable = CanManageExtensions;
                     ToolTip = 'Specifies whether the runtime should allow this extension to make HTTP requests through the HttpClient data type when running in a non-production environment.';
                 }
             }
+
+            group("Resource protection policies")
+            {
+                Visible = IsTenantExtension;
+
+                Caption = 'Resource Protection Policies defined by the extension';
```
Comment (NKarolak) [inline src/System Application/App/Extension Management/src/ExtensionSettings.Page.al]:
```suggestion
                Caption = 'Resource Protection Policies defined by the extension';
                Visible = IsTenantExtension;
                
```

**Review by NKarolak — APPROVED** [review]
This was supposed to be my last review ;-)

**Review by StefanSosic — APPROVED** [review]

**Review by StefanSosic — APPROVED** [review]

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al [hunk 14]
```diff
@@ -208,37 +215,15 @@ page 2500 "Extension Management"
                         CurrPage.Update(false);
                     end;
```
Comment (grobyns) [inline src/System Application/App/Extension Management/src/ExtensionManagement.Page.al:230]: removing the action is a breaking change. You would have to obsolete it, add preprocessor symbols and hide it + make it do nothing...

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al [hunk 15]
```diff
@@ -273,7 +258,7 @@ page 2500 "Extension Management"
             {
                 Caption = 'Manage', Comment = 'Generated from the PromotedActionCategories property index 4.';
```
Comment (grobyns) [inline src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]: removing the actionref is also breaking afaik

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al (hunk 14 above)
Comment (SBalslev) [reply src/System Application/App/Extension Management/src/ExtensionManagement.Page.al:230]: thank you

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al [hunk 16]
```diff
@@ -208,37 +215,33 @@ page 2500 "Extension Management"
...
                 }
+#endif
                 action("Microsoft AppSource Gallery")
                 {
                     Caption = 'AppSource Gallery';
                     Enabled = IsSaaS;
                     Image = NewItem;
                     ToolTip = 'Browse the Microsoft AppSource Gallery for new extensions to install.';
                     Visible = not IsOnPremDisplay;
+                    RunObject = Page 2515; // Using ID to avoid circular dependency.
```
Comment (grobyns) [inline src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]: this creates a logical dependency issue. The way to fix it is to have the AppSource Gallery module add this action using a page extension.

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al [hunk 17]
```diff
@@ -487,12 +489,6 @@ codeunit 2501 "Extension Marketplace"
...
     local procedure OpenAppSourceMarket()
     begin
-#if not CLEAN24        
-#pragma warning disable AL0432
-        Page.Run(Page::"Extension Marketplace");
-#pragma warning restore AL0432
-#else        
-        Hyperlink('https://aka.ms/BCAppSource');
-#endif        
+        Page.Run(2515); // Using ID to avoid circular dependency.
```
Comment (grobyns) [inline src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al]: same... this subscriber should be placed in the other module

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Page.al [hunk 18]
```diff
@@ -32,85 +28,38 @@ page 2502 "Extension Marketplace"
...
+    actions
+    {
+        area(Navigation)
+        {
+            action("Microsoft AppSource Gallery")
+            {
+                Caption = 'AppSource Gallery';
+                Image = NewItem;
+                ToolTip = 'Browse the Microsoft AppSource Gallery for new extensions to install.';
+                RunObject = Page 2515; // Using ID to avoid circular dependency.
```
Comment (grobyns) [inline src/System Application/App/Extension Management/src/ExtensionMarketplace.Page.al]: move to appsource gallery module

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al (hunk 17 above)
Comment (SBalslev) [reply src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al]: Good one - thanks

**Review by grobyns — APPROVED** [review]

File: src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al [hunk 19]
```diff
@@ -43,7 +43,6 @@ codeunit 2501 "Extension Marketplace"
         MarketPlaceUnsuccInstallTxt: Label 'The market place extension installation has failed with the result ''%1''. Error message: ''%2''', Comment = '%1 - OperationResult parameter value, %2 - Error message';
         AlreadyInstalledMsg: Label 'The extension %1 is already installed.', Comment = '%1=name of app';
         AppsourceTxt: Label 'https://appsource.microsoft.com', Locked = true;
```
Comment (blrobl) [inline src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al:46]:
```suggestion
```

Comment (SBalslev) [reply src/System Application/App/Extension Management/src/ExtensionMarketplace.Codeunit.al:46]: missing clean 25

**Review by blrobl — DISMISSED** [review]

File: src/System Application/App/Extension Management/permissions/ExtensionManagementObjects.PermissionSet.al [hunk 20]
```diff
@@ -17,19 +17,16 @@ permissionset 2504 "Extension Management - Objects"
     Permissions = table "Extension Installation" = X,
                   codeunit "Data Out Of Geo. App" = X,
                   codeunit "Extension Management" = X,
+#if not CLEAN25
                   codeunit "Extension Marketplace" = X,
```
Comment (blrobl) [inline src/System Application/App/Extension Management/permissions/ExtensionManagementObjects.PermissionSet.al:21]:
```suggestion
page 
```

**Review by blrobl — APPROVED** [review]

**Review by grobyns — APPROVED** [review]
